src/web/builds.rs

```diff
@@ -37,63 +38,82 @@ impl_webpage! {
 pub fn build_list_handler(req: &mut Request) -> IronResult<Response> {
     let router = extension!(req, Router);
     let name = cexpect!(req, router.find("name"));
-    let version = cexpect!(req, router.find("version"));
+    let req_version = router.find("version");
 
     let mut conn = extension!(req, Pool).get()?;
-    let limits = ctry!(req, Limits::for_crate(&mut conn, name));
 
-    let query = ctry!(
-        req,
-        conn.query(
-            "SELECT crates.name,
-                releases.version,
-                releases.description,
-                releases.rustdoc_status,
-                releases.target_name,
-                builds.id,
-                builds.rustc_version,
-                builds.cratesfyi_version,
-                builds.build_status,
-                builds.build_time
-             FROM builds
-             INNER JOIN releases ON releases.id = builds.rid
-             INNER JOIN crates ON releases.crate_id = crates.id
-             WHERE crates.name = $1 AND releases.version = $2
-             ORDER BY id DESC",
-            &[&name, &version]
-        )
-    );
+    match match_version(&mut conn, name, req_version).and_then(|m| m.assume_exact())? {
```

Comment:
This might be simpler as an early-return to avoid nesting everything

```rust
let version = match match_version(&mut conn, name, req_version).and_then(|m| m.assume_exact())? {
    MatchSemver::Exact((version, _)) => version,
    MatchSemver::Semver((version, _)) => {
        let url = ctry!(
            req,
            Url::parse(&format!(
                "{}/crate/{}/{}/builds",
                redirect_base(req),
                name,
                version
            )),
        );

        return Ok(super::redirect(url));
    }
};
```

Reply:
:facepalm: I like that a lot better. I have updated the PR to use early returns. The diff is much cleaner, now.